packages/block-editor/src/store/selectors.js

```diff
@@ -1537,3 +1539,128 @@ export function didAutomaticChange( state ) {
 export function __unstableGetSelectedMountedBlock( state ) {
 	return state.selectedMountedBlock;
 }
+
+/**
+ * Get the label for the block, usually this is either the block title,
+ * or the value of the attribute denoted by the block's `label` property.
+ *
+ * @param {Object} state    Store state.
+ * @param {string} clientId ClientId for the block.
+ *
+ * @return {string} The block label.
+ */
+export function getBlockLabel( state, clientId ) {
+	const blockName = getBlockName( state, clientId );
+
+	const {
+		__experimentalLabel: labelAttribute,
+		title,
+	} = getBlockType( blockName );
+
+	if ( ! labelAttribute ) {
+		return title;
+	}
+
+	const attributes = getBlockAttributes( state, clientId );
+	const label = attributes[ labelAttribute ];
+
+	if ( ! label ) {
+		return title;
+	}
+
+	// Strip any HTML (i.e. RichText formatting) before returning.
+	return toPlainText( label );
```

Comment:
Following from my comment at https://github.com/WordPress/gutenberg/pull/18132#discussion_r365323192

If we're expecting the accessible label to be associated with an attribute, I still find it strange we'd need to do this, because in my view, if the attribute value can represent a label, it should already be plain text.

I guess I can understand based on what you're saying in https://github.com/WordPress/gutenberg/pull/18132#discussion_r366784819 that this could help enable re-use of attributes where the attribute itself serves some purpose (which may require it to contain markup) but can also represent its title (but only after some post-processing). The example you gave is something like a heading's `title`. I kinda feel like this should still be left to the block though, or at least:

- If we're already going to need to provide some way for a block to define the logic associated with deriving an accessible label from attributes, then this would force some consistency and simplify the interface (i.e. `__experimentalLabel` is not an overloaded block setting)
- What would be the full extent of the "global" post-processing we apply here? I find it troubling that we assume that this value could contain rich text. What else could it contain? My point is that if we force this to be handled by a block, then we don't need to answer this question, because that's left to the block to decide.

The way I'd see it is something more like:

```js
registerBlockType( 'core/heading', {
	attributes: {
		title: {
			type: 'string',
		},
	},
	getLabel: ( { attributes } ) => toPlainText( attributes.title ),
	// ...
} );
```

Reply:
The main issue is formatting, for example the Navigation Link block, which has a label that can be formatted (so could contain tags like `em`/`strong`).

The label is no longer a function, instead you can denote a single attribute:
```
__experimentalLabel: 'label'
```
so there's no way for `toPlainText` to be called from there.

In terms of why it's no longer a function I'll need to defer to @mtias or @youknowriad on this, but I think it probably needs some time to mature or to think about the wider use cases.

Reply:
> The main issue is formatting, for example the Navigation Link block, which has a label that can be formatted (so could contain tags like `em`/`strong`). I didn't want this to appear in the block navigation.

Yeah, this makes sense to me that we wouldn't want these tags to be represented. My concern is more that with the implementation as proposed, it (implicitly) bakes in handling of specific attribute values where a string may or may not contain some HTML. It all seems very loosely-defined. The strings containing HTML is an implementation detail of rich text, and is neither something the block editor should be aware of, nor that it actively considers (we just process all strings, regardless whether they're "rich text" strings). It also raises questions for which I don't think we have a good baseline: What criteria do we have for considering whether it makes sense to bake-in such processing? Should we also strip shortcodes from strings? Should we `Array#join` an array of strings, if that's how a block chooses to represent some attribute value?

I'd also invite some perspective from Matias and Riad on whether we should avoid the function.

My initial impression is that, by delegating this to the block implementer, it may be slightly less convenient, but is much clearer how this relationship is represented. A block implementer knows that they're working with a rich text value, and they would know that the label must be provided as plain text. It's up to them to satisfy that requirement. We could always consider to add the attribute-name form as an abstracted shorthand in the future, if we can answer the clearer picture for how those problematic attribute value forms be interpreted by the block editor.

Reply:
What are the negatives, though? Perhaps a slight overhead in cleaning these strings. 

The positives seem to outweigh the negatives to me—not only is this easier for block implementors, but also the editor won't show an unusual looking html string if a block implementor forgot to convert the text. It feels like it's a very easy thing to miss, but it'd look really buggy to the end user.

Reply:
Would it be strange to serialise the block and strip the HTML from that?

Reply:
After all, this is what a screen reader would read if there were no label. It would read the whole content of the block?

Reply:
> Would it be strange to serialise the block and strip the HTML from that?

> After all, this is what a screen reader would read if there were no label. It would read the whole content of the block?

Most of the points I raised above were related to the block navigation UI:
![Screen Shot 2020-01-21 at 3 12 47 pm](https://user-images.githubusercontent.com/677833/72783102-8db1fb00-3c60-11ea-8557-1cad4fef3ffa.png)

When there's no label it's just the block title that's output. Also worth noting that this is output as a string, so any markup will appear as plain text (e.g. `<em>text</em>` instead of <em>text</em>). It doesn't use `dangerouslySetInnerHTML`

It's the same for the accessiblity label, since this is now implemented in navigation mode, the content of the block wouldn't be read, just the block title. If tags were present it'd be output in the string used as the aria-label:
```
aria-label="<em>text</em>"
```
All of that html would be read "less em greater text less slash em greater".

A block outputs HTML, so not an issue there.

If that can be avoided, why not?